### PR TITLE
[FIRRTL/IMConstantProp] Make field sensitive

### PR DESCRIPTION
**Pull request by uenoku**
This PR changes IMCostantProp field sensitive. The main difference is to use `pair<Value, unsigned>` instead of `Value` as keys of lattice values. The second element represents the position of the value in **leafs** which is different from firrtl field index because field index is the position at tree. 

For example, for `bundle<a: uint<1>, b: vector<uint<1>, 5>, c: uint<1>>`,  leaf Indices of a, b and c are 0, 1, 6 respectively.

### CODE REVIEW

**Comment by lattner** [comment]
Nice, this is looking like the right path!

**Comment by uenoku** [comment]
It's ready for review. I implement `ValueAndLeafIndex` class to use the key to associate with lattice values. Main concern is that there are bunch of caches related to calculate the leaf index. 

There are still debugging messages but I'd like to remove later of the review.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 1]
```diff
@@ -9,25 +9,60 @@
...
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
 #include "circt/Dialect/FIRRTL/InstanceGraph.h"
 #include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/LLVM.h"
 #include "mlir/IR/Threading.h"
 #include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/TinyPtrVector.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:21]: I think that some of these headers are already transitively included, please prune out what isn't necessary

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 2]
```diff
@@ -9,25 +9,60 @@
...
 
 using namespace circt;
 using namespace firrtl;
 
-/// Return true if this is a wire or register.
+/// Return true if this is either wire, register or similar(subfield and
+/// subindex).
 static bool isWireOrReg(Operation *op) {
-  return isa<WireOp>(op) || isa<RegResetOp>(op) || isa<RegOp>(op);
+  return isa<WireOp, RegResetOp, RegOp, SubfieldOp, SubindexOp>(op);
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:31]: Why add subfield/subindex to this predicate?  Should the function be renamed?

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 3]
```diff
@@ -9,25 +9,60 @@
...
 
+/// Return true if a type is possible to track. Currently, we allow passive
+/// types.
+static bool isTrackableType(Type type) {
+  return type.cast<FIRRTLType>().isPassive();
+}
+
+/// This function recursively apply `fn` to leaf ground types of `type`.
+static void traverseFIRRTLType(Type type,
+                               llvm::unique_function<void(FIRRTLType)> fn) {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:47]: I think this should move up to FIRRTLTypes

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 4]
```diff
@@ -151,14 +186,74 @@ class LatticeValue {
...
+
+  // isKnownRoot is true if we already know that the value is a root value.
+  // If this flag is true, we can skip the translation.
+  bool isKnownRoot() const { return valueAndFlag.getInt(); }
+  void setKnownRoot() { valueAndFlag.setInt(true); }
+
+private:
+  // Pair of a value and flag.
+  llvm::PointerIntPair<Value, 1, bool> valueAndFlag{};
+  unsigned leafId{};
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:224]: nit, but please use `= 0` isntad of {}.  You can drop the {} from valueAndFlag

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 5]
```diff
@@ -151,14 +186,74 @@ class LatticeValue {
...
+public:
+  ValueAndLeafIndex(Value value, unsigned leafId, bool isKnownRoot = false)
+      : valueAndFlag(value, isKnownRoot), leafId(leafId) {}
+  ValueAndLeafIndex() = default;
+  Value getValue() const { return valueAndFlag.getPointer(); }
+  unsigned getLeafIndex() const { return leafId; }
+
+  // isKnownRoot is true if we already know that the value is a root value.
+  // If this flag is true, we can skip the translation.
+  bool isKnownRoot() const { return valueAndFlag.getInt(); }
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:218]: What does the word "known" convey?  Can this be "isRoot"?

Also, is the terminology correct here, is this really a leaf or a root?    There can only be one root but many leaves.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 6]
```diff
@@ -169,41 +264,109 @@ struct IMConstPropPass : public IMConstPropBase<IMConstPropPass> {
...
     return it != latticeValues.end() && it->second.isOverdefined();
   }
 
-  /// Mark the given value as overdefined. This means that we cannot refine a
-  /// specific constant for this value.
-  void markOverdefined(Value value) {
+  // Value is regarded as overdefined if one of lattice values associated with
+  // the value is overdefined.
+  bool isOverdefined(Value value) const {
+    auto [root, index, size] =
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:275]: Is this a C++'20 thing?  We use C++'17, but I haven't seen us use this extension before.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 7]
```diff
@@ -383,50 +700,123 @@ void IMConstPropPass::markBlockExecutable(Block *block) {
...
+                      InvalidValueAttr::get(destType));
+  }
 }
 
 void IMConstPropPass::markRegResetOp(RegResetOp regReset) {
-  // If the reg has a non-ground type, then it is too complex for us to handle,
-  // mark it as overdefined.
-  // TODO: Eventually add a field-sensitive model.
-  if (!regReset.getType().getPassiveType().isGround())
+  // If the reg has a passive type, then it is too complex for us to
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:728]: `If the reg has a NON passive type,`. ==> is this actually allowed in the firrtl dialect?  What is an example?

**Review by lattner — COMMENTED** [review]
This is looking great, another pass of comments, thanks!

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 7 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:728]: Registers have to be passive. This check is probably unnecessary and you can just remove it in a push to `main`.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 8]
```diff
@@ -383,50 +700,123 @@ void IMConstPropPass::markBlockExecutable(Block *block) {
...
+  if (!isTrackableType(regReset.getType()))
     return markOverdefined(regReset);
 
   // The reset value may be known - if so, merge it in.
-  auto srcValue = getExtendedLatticeValue(regReset.resetValue(),
-                                          regReset.getType().cast<FIRRTLType>(),
+  auto &destTypes = getLeafGroundTypes(regReset.getType());
+
+  // Iterate over each ground type and merge lattice values separatetly.
+  for (unsigned i = 0, e = destTypes.size(); i < e; i++) {
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:737]: Nit: the convention in CIRCT tends to be towards always using preincrement, even though it doesn't matter here. Ditto for other uses here.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 9]
```diff
@@ -383,50 +700,123 @@ void IMConstPropPass::markBlockExecutable(Block *block) {
...
+
+  // Register the current position to `valueToRootValueAndLeafIndex`.
+  valueToRootValueAndLeafIndex[subfield] = rootAndIndex;
+
+  for (unsigned i = 0; i < getNumberOfLeafs(subfield.getType()); i++)
+    rootToChildrenAccess[{rootAndIndex.getValue(),
+                          rootAndIndex.getLeafIndex() + i,
+                          /*isKnownRoot=*/true}]
+        .push_back(subfield);
+}
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:805]: These methods could be combined to save on the duplication (with an internal switch to choose between `.fieldIndex()` and `.index()`.

File: test/Dialect/FIRRTL/imconstprop.mlir [hunk 10]
```diff
@@ -542,3 +542,91 @@ firrtl.circuit "dntOutput" {
...
+    firrtl.connect %1, %c1_ui1 : !firrtl.uint<1>, !firrtl.uint<1>
+    %2 = firrtl.xor %0, %1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
+    firrtl.connect %b, %2 : !firrtl.uint<1>, !firrtl.uint<1>
+    // CHECK:      %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+    // CHECK-NEXT: firrtl.connect %b, %c0_ui1 : !firrtl.uint<1>, !firrtl.uint<1>
+  }
+}
+
+firrtl.circuit "VectorPropagation2" {
+  firrtl.module @VectorPropagation2(in %clock: !firrtl.clock, out %b1: !firrtl.uint<6>, out %b2: !firrtl.uint<6>, out %b3: !firrtl.uint<6>) {
```
Comment (seldridge) [inline test/Dialect/FIRRTL/imconstprop.mlir:564]: Most tests in here are split up (`// -----`).  It would be good to continue that convention.

Also, these tests look great.  For completeness, can you add a test that includes a bundle of vectors?

**Review by seldridge — COMMENTED** [review]
I need to give this one more look, but this is looking solid.

File: test/Dialect/FIRRTL/imconstprop.mlir (hunk 10 above)
Comment (uenoku) [reply test/Dialect/FIRRTL/imconstprop.mlir:564]: Thanks, I will add tests with a bundle of vectors, and also asyncreset and nested bundles.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 11]
```diff
@@ -151,14 +186,74 @@ class LatticeValue {
...
+  return os;
+}
+
+// This class defines the data structure associated with lattice values.
+// This means basically a pair of value and integer. The integer
+// represents the relative offset to the value on leaf level of type.
+class ValueAndLeafIndex {
+public:
+  ValueAndLeafIndex(Value value, unsigned leafId, bool isKnownRoot = false)
+      : valueAndFlag(value, isKnownRoot), leafId(leafId) {}
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:211]: `isKnownRoot` encoding is used to avoid recomputation of the root value. But on the second thought, I think it is sufficient to check whether `isa<SubfieldOp, SubindexOp>` is true.

> I think this is equivalent to leafId == 0

`leafId==0` doesn't always indicate the value is root. For example,  
```
%a = reg : firrtl.vector<vector<uint, 3>, 3>
%a_0 = firrtl.subindex %a[0]
%a_0_0 = firrtl.subindex %a_0[0]
```
LeafIds of `%a, %a_0, %a_0_0` are all 0 because they point at first element. However, only `%a` must be root.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 5 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:218]: > What does the word "known" convey? Can this be "isRoot"?
Yeah, it's confusing so I'll change to "isRoot". I used isKnownRoot to avoid recomputation of the root value. 
I thought it's necessary to lookup a map to check whether the value is root or not.
But on the second thought, I think it is sufficient to check whether `isa<SubfieldOp, SubindexOp>` is true.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:31]: Thanks, I thought it was convenient to regards subfield/subindex as wire/reg at post processing. But it's confusing so I'll separate it.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 7 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:728]: Thank you, it was my misunderstanding.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 3 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp]: Thanks!

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 8 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp]: Thanks!

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 9 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp]: Thanks, I combined them

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 6 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:282]: Yes, I prefer structured bindings over `std::tie` :)

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 4 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp]: Thanks!

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 12]
```diff
@@ -151,14 +193,74 @@ class LatticeValue {
...
   llvm::PointerIntPair<Attribute, 2, Kind> valueAndTag;
 };
+
+raw_ostream &operator<<(raw_ostream &os, const LatticeValue &lattice) {
+  os << "<unknown: " << lattice.isUnknown()
+     << ", invalid: " << lattice.isInvalidValue()
+     << ", constant: " << lattice.isConstant()
+     << ", over: " << lattice.isOverdefined() << ">";
+  return os;
+}
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:210]: Can more than one of these ever be true? I think not? That would then motivate just having one serialization to one string, e.g., `<unknown`> or `<invalid>`.

**Comment by uenoku** [comment]
I found that a connection between aggregate types doesn't propagate constant values properly. I'll let you know when that's fixed.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 6 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:282]: We might have had problems with this feature in windows.  You should check MSVC.

**Comment by darthscsi** [comment]
> I found that a connection between aggregate types doesn't propagate constant values properly. I'll let you know when that's fixed.

It's ok to do this in several steps.  I.e. first the framework with maybe one op constant propping, then make other operations field sensitive.  This may simplify things.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 6 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:282]: Thanks, I ran windows ci but it passed so I assume it's fine.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 7 above)
Comment (lattner) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:725]: Plz make it an assert though!

**Comment by uenoku** [comment]
> It's ok to do this in several steps. I.e. first the framework with maybe one op constant propping, then make other operations field sensitive. This may simplify things.

Thanks. That makes sense. I might do so if I can’t figure the cause of issue easily.

**Comment by uenoku** [comment]
I changed not to remove aggregate register/wire/connections even though they are all constant. 
It's a problematic when aggregates are used as source values. Fox example, 
```scala
circuit Foo:
  module Foo:
    input clock : Clock
    input reset : UInt<1>
    output res0: UInt<3>
    output res1: UInt<3>

    wire init : UInt<3>[2] 
    reg reg : UInt<3>[2] , clock with : ( reset => ( reset , init ) )
    init[0] <= UInt<3>(0)
    init[1] <= UInt<3>(2)
    reg[0] <= UInt<3>(1)
    reg[1] <= UInt<3>(2)
    res0 <= reg[0]
    res1 <= reg[1]
```
In this example, `init` is an constant array `[0, 2]` but it used as reset value of `reg`. `res0` is not constant folded because `res0` is either 0 or 1. Therefore, we can't remove `reg` because it's ued as source value of res0.

------

To summarize the current status of this PR, 

Scope:
1. Propagate constant through passive type including aggregates
2. Replace ground type element if possible

Not in scope:
1. Deletion of aggregate register/wire/connections. A more complicated analysis would be necessary, i.e. we need to check all uses of subelements and ensure that it's safe to delete.   
2. Propagate through non-passive type

**Comment by uenoku** [comment]
On the second thought, it might be better to use `FieldRef` instead of adding new class `ValueAndLeafIndex`...
When I first implemented this, I thought using leaf id is better idea because we can iterate over consecutive leafs with indices. 
However, now I don't think there is much difference. For example, `translateAndFind` is almost same to `getFieldRefFromValue`.

**Comment by uenoku** [comment]
Talked to @seldridge and we came to the conclusion that `FieldRef` would be better. So I would like to refactor this PR to use FieldRef.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 13]
```diff
@@ -170,109 +241,265 @@ struct IMConstPropPass : public IMConstPropBase<IMConstPropPass> {
...
+
+  /// Return a vector of pair of fieldIDs and ground types which `type` has.
+  const SmallVector<std::pair<unsigned, FIRRTLType>> &
+  getLeafFieldIDsAndGroundTypes(Type type) const {
+    auto &entry = typeToLeafFieldIDsAndGroundTypes[type];
+    if (!entry.empty())
+      return entry;
+    auto fn = [&entry](unsigned fieldID, FIRRTLType type) {
+      entry.push_back({fieldID, type});
+    };
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:453]: Nit: you can inline this if you want.

File: test/Dialect/FIRRTL/imconstprop.mlir [hunk 14]
```diff
@@ -542,3 +542,198 @@ firrtl.circuit "dntOutput" {
...
+    // CHECK: %bundle_vec = firrtl.reg %clock
+    // CHECK: firrtl.subindex %bundle_vec[0] : !firrtl.vector<bundle<a: uint<3>, b: uint<3>>, 3>
+    // CHECK: firrtl.subindex %bundle_vec[2] : !firrtl.vector<bundle<a: uint<3>, b: uint<3>>, 3>
+
+    // Check that results are constant folded.
+    // CHECK:      %[[c4_ui3:.+]] = firrtl.constant 4 : !firrtl.uint<3>
+    // CHECK:      %[[c5_ui3:.+]] = firrtl.constant 5 : !firrtl.uint<3>
+
+    // CHECK:      firrtl.connect %res1, %[[c4_ui3]] : !firrtl.uint<3>, !firrtl.uint<3>
+    // CHECK-NEXT: firrtl.connect %res2, %[[c5_ui3]] : !firrtl.uint<3>, !firrtl.uint<3>
```
Comment (seldridge) [inline test/Dialect/FIRRTL/imconstprop.mlir:690]: This type of stuff would be better to put in an XFAIL test.

**Review by seldridge — COMMENTED** [review]
This is looking great and super clean with the use of fieldrefs. Nice work.

Can you add some tests that hit `regreset` and that also do various combinations of don't touch annotations.  This latter part has been a big paint point for us and I want to make sure there aren't any issues there.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 15]
```diff
@@ -24,16 +28,69 @@ static bool isWireOrReg(Operation *op) {
...
+/// Return true if a type is possible to track. Currently, we allow passive
+/// types.
+static bool isTrackableType(Type type) {
+  return type.cast<FIRRTLType>().isPassive();
+}
+
+/// This function recursively applies `fn` to leaf ground types of `type`.
+static void
+foreachFIRRTLGroundType(FIRRTLType type,
+                        llvm::unique_function<void(unsigned, FIRRTLType)> fn) {
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:57]: Yes, it makes totally sense to me. There are really bunch of this kind of functions so it would be really useful.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 16]
```diff
@@ -24,16 +28,69 @@ static bool isWireOrReg(Operation *op) {
...
 //===----------------------------------------------------------------------===//
 
 namespace {
+
+// Printer for FieldRef.
+raw_ostream &operator<<(raw_ostream &os, const FieldRef &fieldRef) {
+  os << getFieldName(fieldRef);
+  return os;
+}
+
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp]: Thanks. I moved this into FIRRTLDialect.{h, cpp}

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 17]
```diff
@@ -170,109 +241,265 @@ struct IMConstPropPass : public IMConstPropBase<IMConstPropPass> {
...
   /// Merge information from the 'from' lattice value into value.  If it
   /// changes, then users of the value are added to the worklist for
   /// revisitation.
-  void mergeLatticeValue(Value value, LatticeValue &valueEntry,
+  void mergeLatticeValue(FieldRef fieldRef, LatticeValue &valueEntry,
                          LatticeValue source) {
-    if (!source.isOverdefined() && hasDontTouch(value))
+    assert(isRoot(fieldRef) && "value must be known to be root beforehand");
+    if (!source.isOverdefined() && hasDontTouch(fieldRef.getValue()))
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:273]: Yeah, it is not field sensitive now so I agree that this is too restrictive.  I will make it only block the field with don't touch.

File: test/Dialect/FIRRTL/imconstprop.mlir (hunk 14 above)
Comment (uenoku) [reply test/Dialect/FIRRTL/imconstprop.mlir:690]: I'm not sure I understand correctly but it doesn't contain failure tests. Could you elaborate?

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 18]
```diff
@@ -170,109 +241,265 @@ struct IMConstPropPass : public IMConstPropBase<IMConstPropPass> {
...
+  // Value is regarded as overdefined if one of lattice values associated with
+  // the value is overdefined.
+  bool isOverdefined(Value value) const {
+    auto [root, index, size] = getRootValueWithCorrespondingLeafIDRange(value);
+    auto types = getLeafFieldIDsAndGroundTypes(root.getType());
+    for (unsigned i = 0; i < size; ++i)
+      if (isOverdefined({root, types[index + i].first}))
+        return true;
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:243]: Yes, that's correct. Currently this function will not be applied to aggregate values. I'll delete it. Thanks!

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp]: I added a function `bool hasDontTouch(FieldRef fieldRef)` to check the annotation in the field sensitive way. 
But I haven't actually implemented it for now. Because we have to check the donttouch doesn't exist any of parent of given fieldID, but  the current Annotation class doesn't have the utility to check that one fieldID is a parent of another fieldID. It will require moderate amount of code so I'd like to do it in another patch. Therefore, I made it very conservatively, just return whether root has donttouch or not.

**Review by seldridge — APPROVED** [review]
This looks great.

Before you merge this, can you do two things:

1. Rebase and fix conflicts
2. Run this on some cores, with lowering and IMCP on, and make sure the circuit is _exactly_ the same as without this commit.

(2) is really important because I don't want to introduce tedious IMCP regressions...

Thanks for working so hard on getting this together, Hideto!

**Comment by uenoku** [comment]
> Rebase and fix conflicts
Run this on some cores, with lowering and IMCP on, and make sure the circuit is exactly the same as without this commit.

Yes, I confirmed that we can get the same results for several cores. But I noticed the regression is more than expected; it almost doubles the execution time of IMCP so I'll try to reduce them before land this.

**Comment by uenoku** [comment]
Sorry for the very long delay. 
* I refactored to remove leafID at all since it can be just replaced by fieldID. I bet the performance regression comes from there so I guess the current performance is tolerbale (though there should be some regressions).
* There were basically bugs around lattice states updates around aggregate ports and fold results. I ported some of normal imcp tests by just promoting their values into aggregate values.

File: lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp [hunk 19]
```diff
@@ -198,8 +198,15 @@ bool AnnotationSet::hasAnnotationImpl(StringRef className) const {
...
-  return hasAnnotation(dontTouchAnnoClass);
+bool AnnotationSet::hasDontTouch(unsigned fieldID) const {
+  if (!hasAnnotation(dontTouchAnnoClass))
+    return false;
+  if (fieldID == 0)
+    return true;
+
+  // FIXME: We need to check the fieldID of dontTouchAnnotation is a parent of
+  // `fieldID` here. Currently, we handle it conservatively.
+  return true;
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp]: We now use symbols as donTouch so it gets harder to reason in the field sensitive way..

File: test/Dialect/FIRRTL/imconstprop-aggregate.mlir [hunk 20]
```diff
@@ -0,0 +1,447 @@
...
+firrtl.circuit "VectorPropagation1" {
+  // CHECK-LABEL: @VectorPropagation1
+  firrtl.module @VectorPropagation1(in %clock: !firrtl.clock, out %b: !firrtl.uint<1>) {
+    %c1_ui1 = firrtl.constant 1 : !firrtl.uint<1>
+    %tmp = firrtl.reg %clock  : !firrtl.vector<uint<1>, 2>
+    %0 = firrtl.subindex %tmp[0] : !firrtl.vector<uint<1>, 2>
+    firrtl.connect %0, %c1_ui1 : !firrtl.uint<1>, !firrtl.uint<1>
+    %1 = firrtl.subindex %tmp[1] : !firrtl.vector<uint<1>, 2>
+    firrtl.connect %1, %c1_ui1 : !firrtl.uint<1>, !firrtl.uint<1>
+    %2 = firrtl.xor %0, %1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
```
Comment (darthscsi) [inline test/Dialect/FIRRTL/imconstprop-aggregate.mlir:12]: both the read and the write derive from the same subindex op.  A more thorough test would have them derive from different subindex ops so that you were sure the lattice was propagated correctly through the connect.

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 21]
```diff
@@ -363,6 +501,29 @@ LatticeValue IMConstPropPass::getExtendedLatticeValue(Value value,
...
+    // record the dependency.
+    if (llvm::all_of(result.getUsers(), [](Operation *op) {
+          return isa<SubaccessOp, SubfieldOp>(op);
+        }))
+      return;
+
+    foreachFIRRTLGroundType(
+        result.getType().cast<FIRRTLType>(),
+        [&](unsigned relativeFieldID, auto) {
+          fieldRefToSubelementValues[fieldRef.getSubField(relativeFieldID)]
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:518]: I prepared this map to avoid unnecessary propagations between subWhatever ops. This implementation doesn't require propagation between subWhatever ops and their parent ops since the states are shared as FieldRef.

File: test/Dialect/FIRRTL/imconstprop-aggregate.mlir (hunk 20 above)
Comment (uenoku) [reply test/Dialect/FIRRTL/imconstprop-aggregate.mlir:12]: That makes sense. Thanks.

**Comment by uenoku** [comment]
Close the PR for now.
